### PR TITLE
modulartiy_max: introduce enforce_n_communities parameter

### PR DESCRIPTION
**Pull request by martacki**
In modularity_max, I introduced the additional parameter ``n_communitites`` in  #4950.
However, for my application I want to sometimes go below the optimal community structure and continue the merging process until n_communities is reached, regardless if this does not further improve modulartiy. In that case I'd want to merge those nodes that decrease modularity the least. I propose a solution to solve this issue in this PR.

2 other options to implement this without the `enforce_n_communities` parameter:
1. Always proceed this way if n_communities is not 1. This has downside when n_communities indeed should be 1, but optimal modularity is archieved for a larger network. This is common in my application, but theoretically could be circumvent in the specific use-case. (don't like this option)
2. n_communities is `None` by default, and if given, the number of communities is always enforced (i.e. also if n_communities=1; might be nicer than current suggestion).

### CODE REVIEW

**Comment by dschult** [comment]
This makes me consider again a generator interface. There are so many possibilities (including sorted and not sorted) that the flexibility of a generator function could serve all these cases.  We decided not to use a generator in the previous PR for backwards compatibility.  But this suggestion makes me think of the following approach:

make a new function that iterates through the communities.  Have `modularity_max` retain its original signature and use the iterator to construct its return value.  That way we get the iterator functionality through the generator interface but we keep the old interface for ease of use and backward compatibility.

thoughts?

**Comment by martacki** [comment]
First of all: sorry for the late reply...

I just played a bit with yielding the newly merged nodes/sets instead of returning the sorted list of clusters, but to me the intermediate results are somewhat confusing. I'm not sure how you imagine a sorted generator object, because the sorting of clusters is conducted after the whole list is created. So one can't yield a complete cluster where the merging process is  "finished" so to say; you might get stuck with a subset of your cluster for a while where only a subset of nodes is in compared to the "best" cluster. The remaining nodes in such a case would be added in a (much) later iteration. There will be better merging options in between for other nodes. However this is only true if `only the merged cluster` is yielded.

What I could imagine to work is to always yield _all_ the currently existing clusters, so in the first iteration that would be a full list of each node forming its own cluster. This makes more sense to me.

If you meant something else, could you please specify? But if the latter is what you meant, I could push a suggestion if you want.

**Comment by dschult** [comment]
What I intended is to yield *all* the currently existing clusters. This behaves like a dendogram with each node in its own cluster at the beginning, and then with each iteration you see the clusters after the latest merge.
:}

**Comment by martacki** [comment]
Aha, that's basically what I meant in the second part. Thanks for clarification! I can come up with a draft tomorrow.

Suggestions on the name of the new generator object function? `greedy_modularity_communities_generator(G, weight, resolution)`? (I'd drop `n_communities` for the generator then, as you already suggested in a previous discussion. `n_communities` could then remain in the "original" `greedy_modularity_communities` that'll use the generator to construct the output. I'll add the new `enforce_n_communities` there, too - or will you deprecate the non-generator function in a future release?)

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:17]: These two keyword arguments `n_communities` and `enforce_n_communities` seem like there ought to be a better naming scheme.  The `n_communities` is a cutoff after which we stop. The `enforce_n_communities` is a signal to keep going until n is reached.  I think what I don't like is that `n_communities` means two different things depending on what another keyword arg's value is.

I'd like to propose that we use `best_n_communities=None` to be number n when we want exactly n communities.
And `n_communities` continue to be the cutoff number n of communities at which to stop if the optimum has not been reached.  A better term might be `cutoff`.  That would require a deprecation, so I'm fine either way.  I'm open to other suggestions too...  

Thoughts?

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:17]: yup, sounds more consistent than the current naming scheme. Will use `cutoff` and `best_n_communities`.

**Comment by martacki** [comment]
ah, finally all checks have passed... sorry for this commit-mess ;) I'm not very consequent about code format and style

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:306]: It seems like this loop should stop based on `communities_nxt` instead of `communities_prv`.
Aren't you yielding one more set of communities than you need?  
I think the other way you also don't have to construct both `nxt` and `prv` before entering the loop.
(I also shorten names here.  See which you prefer. Also I think the argument name might be better
as `best_n` than `best_n_communities` because we expect this computation to be about the communities.)
It's something like: 
```python
    if cutoff is None:
        cutoff = 1
    if best_n is None: 
        best_n = len(G)
    compute `nxt` and `mod_nxt` and set dq=0 to start??
    while len(nxt) > 1 and not (dq > 0 and len(nxt) <= cutoff) and not (dq < 0 and len(nxt) <= best_n):
        prv, mod_prv = nxt, mod_nxt
        construct nxt and mod_nxt
        dq = mod_nxt - mod_prv
    return sorted(nxt, key=len, reverse=True)
```
Much of this is a matter of taste of course, but if it is more readable, that's important.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:256]:
```suggestion
    best_n : int or None, optional (default=None)
        Force community merging to continue until `best_n` communities remain
        even if the modularity decreases. 
        If ``None``, don't force it to continue beyond a maximum.
```

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:250]:
```suggestion
        The number of communities in the partition at which the merging process
        will stop even if a maximum is not yet reached.  If None, continue until
        a maximum modularity is reached or all nodes are in a single community.
```

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:229]: Remove blank line before doc_string

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:111]: Does this sort add significantly to the time? I would think it could be a slow step.
The user can sort the yielded communities by length if needed -- and not if they don't want to spend the cpu time doing that.  

Maybe:  `yield list(communities.values()`

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:114]:
```suggestion
    # Merge the two communities that most increase or least decrease modularity
```

**Review by dschult — COMMENTED** [review]
This looks nice!  I've got some rewording and style suggestions. Feel free to tweak them -- the goal is to make the code and docs as readable as possible.

The only big issue is whether to sort each partition by length of the community sizes. That seems costly. The calling function can sort if they want to, but they don't have to spend that time.  Am I missing something with this suggestion?

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:272]: Hmmmm, the `n_communities` kwarg was always integer, never none. Default setting was `n_communities=1` because that kept the original functionality of the function. So only ways to check if `n_communities` was explicitly set by a user would be using the ** option, or not? If not specified explicitly, one wouldn't notice if the default setting is `n_communities=1` or `cutoff=None`. Not sure though.

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:298]: I think not. Because, if `dq<0`, then we want to return `prv`, not `nxt`. In this case `nxt` has already a worse community structure than `prv`, therefore we want to discard it (unless it is specified in best_n to continue merging). Therefore, `prv` should be defined before the loop.
If you do not want to define it before the loop, what you can do, is define only `nxt` before the loop, but still always return `prv`. However, that case, you'd have to make a distinction after the loop, if `len(prv) = 2` (because here now `len(nxt)=1` and the loop automatically breaks) and then check if `nxt` or `prv` should be returned based on dq.

Not sure if I'm overseeing something

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py]: adapted the rest, though

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:271]: I agree.... we should not check if `n_communities is not None`. Sorry about that.
But setting cutoff to 1 is the same as setting cutoff to None. Right?  The process continues until there is only 1 community in either case. So, maybe we just make the default values `cutoff=1` and `n_communities=1` and check if they are not 1 instead of not None.  ???

Something like: (after changing default `cutoff` value to 1)
```python
if n_communities != 1:
    if cutoff == 1:
        cutoff = n_communities
    else:
        raise error # both cannot be set
```
Does this cover all cases?

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:300]: I see what you are saying about wanting to return `prv` when `dq<0`.  That makes this more complicated.
I'm still having trouble seeing the logic all the way through -- so I know that if I look at this in 6 months I won't see it. :}

Why is it an advantage to split the conditions for stopping into line 299 instead of in the while statement?
I'm also having trouble with why `len(prv) <= cutoff` is the right inequality here?  Maybe I just need to come back later when my brain has reset. :}

Are each of the stopping conditions contained in a test to make sure we've got all cases covered? If so, maybe I just need to look at the tests.
Thanks!

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:300]: I agree it's a bit confusing. The main logic is that in the generator, you can break the while loop before merging two new communities as only the difference `dq` can be calculated. Here, we generate the communities in any case and then check later if the newly constructed communities `nxt` are better or worse. If better, we keep going, if worse we return the better community set `prv`. I can add a comment maybe, so this will become clearer?

Resp ln 299: You can add it in the while statement, but I thought this way is more comprehensive. Adding it to the while statement would include then in total 4 conditions. The first one `if True` and the remaining three `if not False`.
But, as you mentioned before, that's taste. I don't really care and can include all of them in the while statement.

True, it should be an equality, ie. `==,` not `<=`. That's somewhat the same because in each iteration we're just reducing `len(prv)` and initially it has to be `>=` (otherwise we'd end up with an error [here](https://github.com/networkx/networkx/pull/5227/files#diff-a9c150aa8f828e5c4485350f618c6a50626ab1d2463ef0ce84792ab06763bf97R258-R261)). Nevertheless, should be changed to `==`. Good catch!

Yup, all parameters should have tests.
The new kwargs have tests:
`cutoff`: test_cutoff_parameter
`n_best`: test_n_best()
Tests for the other parameters were not changed here (as we expect the same result with the new function). They are covered in the tests, too.

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:271]: I partly agree with setting `cutoff` to default value 1 because on the one hand it's somewhat cleaner, but on the other hand that's inconsistent with `best_n=None` (should be `G.number_of_nodes()` by default, but this doesn't work).

For the latter; if `n_communities` is not given, your suggestion currently would result in an error. We could specify it in the kwargs `n_communities=1` as default (as it is currently), but then I don't know how then to distinguish if it was set by a user or if it just the default setting of the function. I mean, a user could also have specified `n_communities=1`, even though that's default. But someone who does not use `n_communities` should not be punished with a deprecation warning, is my feeling...

Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:271]: I agree with your default value of `n_communities=None`. I had forgotten about the deprecation tests already. :{
I think we should use default values of `cutoff=1, best_n=None`.  They should not be the same type of default because `best_n` indicates two pieces of information: whether to continue until there are n communities (None vs a number) *and* the value of *n* (the number -- but only if we using this stopping method).  

I think we should have keywords: `n_communities=None, cutoff=1, best_n=None`. We can then check setting of  `n_communities` by checking whether it is `None` or not, `cutoff` works without preprocessing once deprecation is handled, and `best_n` hold two pieces of information in a single keyword. 

I think this means your fix for my suggested code is:
```python
if n_communities is not None:
    create a deprecation warning
    if cutoff == 1:
        cutoff = n_communities
    else:
        raise error # both cannot be set
```
Am I understanding correctly?

Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:271]: Another issue to deal with:  For the case when `best_n` is a smaller number than `cutoff`, what do we do?

`cutoff` means we stop. But `best_n` says we have to continue.  I guess we should raise an error in this case too.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:300]: Ack -- OK now I am understanding how the generator and the original functions differ. I think this method means we are recomputing the modularity at every step of the merge.  That un-does all the work that the original function does to compute and track `dq` instead of computing the modularity. The complexity of the resulting algorithm is much worse that the previous function.

We have to make use of the `dq` computed in the generator so we can avoid any modularity computation in the outer function. Does this make sense? 

I'm also trying to save the construction of `nxt` until we know whether it is better than `prv`. The old code did that by looking at `dq` for merging (u, v) before actually merging (u,v). 

I've been trying to find a way to get the generator to output `dq` and the communities. But that also requires constructing the new communities before the outer function knows if it needs them.  Here's a crazy suggestion:
We could have the generator yield `dq` values alternating with the partition `communities.values()`. It first yields the partition with each node in isolation. Then it yields the `dq` value of the next step. Then it yields the partition after that step. Then the `dq` for the next step, and it continues until it returns one community with all the nodes.  This means we have two yield statements inside the generator's `while len(H) > 1:` loop. One right after computing `dq = -negdq`. The other just after merging the `communities`. That way we never compute more in the generator than we need to.

The logic in the outer function seems to make more sense to me then too.  Something like:
```python
    stop_at_max = best_n is None
    if not stop_at_max:
        cutoff = best_n

    comm = next(communities_gen)
    while len(comm) > cutoff:   # stop computing when len(comm) == cutoff
        dq, u, v = next(communities_gen)
        if stop_at_max and dq < 0:    # stop at the max_mod, unless best_n gives a size to go to
            break
        comm = next(communities_gen)
    return sorted(comm, key=len, reverse=True)
```

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:300]: I understand your suggestion and it does make sense, but indeed somewhat crazy! I'm not sure if that makes using the generator outside the new function favorable, as the alternating outputs might be somewhat confusing. Would require careful re-writing the whole documentation in the beginning of the function. But I see your point. How expensive is the calculation of modularity? And how expensive is understanding altering outputs of a generator? Guess that's much harder to put a price on...

Anyways, I can try and play a bit with the code and come up with a draft later today.

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:271]: I think this can not catch the case if `cutoff=1` is set by a user (ie _not_ the default setting), while they set `n_communities=N` to some (other) number N. In this case `n_communities` is even the stronger kwarg than `cutoff`, which it shouldn't be as it is to be deprecated. So, then my suggestion are to either
(i) keep the default setting `cutoff=None` _and_ change the default setting `n_communities=None` (before was 1) - which will make the code more verboose, or
(ii) change default settings to `cutoff=1` _and_ remove `n_communities` from kwargs in favor of **aliases.

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:271]: Agree with your 2nd comment, will raise an error in this case.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:300]: Unfortunately, calculating modularity is very costly -- I don't recall what the complexity is, but basically computing modularity requires all edges of the graph while dq requires just looking at the neighbors of u and v. It's Ok for small graphs, but increasing quickly for bigger graphs.

To make the generator user-friendly we might need to do something like make it private (add underscore to the front of its name and take it out of `__all__`) and then make a front-end generator that yields just the communities. Both the front-end generator and the greedy function would call the private generator.  A few paragraphs of docs explaining the private generator output (in the private generator's docstring will be needed for sure.

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py]: Didn't construct a front-end generator yet. Let's first finish these two.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:302]: Why should the condition involving `best_n` involve `dq` at all.  The doc_string says `best_n` makes the algorithm stop when you have n communities "even if modularity decreases". But the code implements "only if modularity has started to decrease.       

How does the user make it stop at exact n communities?

It seems like we've got too many different possibilities here. You can stop:
1) when modularity is at a max, 
2) when n == cutoff and no max found
3) when n == best_n and max found
[other possibilities]
4) when n == best_n independent of whether max has been found.

Is there a simple way to do this?

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [reply networkx/algorithms/community/modularity_max.py:281]: Can we replace the `aliases` keyword argument with arguments `n_communities` (in the same position it was in before) and `cutoff`?

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:280]: Slightly shorter code in the formatting strings allows `black` to play nice.
```suggestion
    if (cutoff < 1) or (cutoff > G.number_of_nodes()):
        raise ValueError(f"best_n must be between 1 and {len(G)}. Got {best_n}.)"
    if best_n is not None:
        if (best_n < 1) or (best_n > G.number_of_nodes()):
            raise ValueError(f"best_n must be between 1 and {len(G)}. Got {best_n}.")
        if best_n < cutoff:
            raise ValueError(f"Must have best_n >= cutoff. Got {best_n} < {cutoff}")
    else:
        best_n = G.number_of_nodes()
```

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:218]: Another minor detail.  
Is it better to yield a list (causing many probably unneeded lists to be formed) or to yield the values object? 
If the caller wants a list they can make one. If they don't want a list, they can't change this line.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:135]: Minor detail, but shouldn't we just yield this after line 117?  Messing with `dq_heap` isn't needed by the calling function for this `yield`.

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py]: done. can we resolve this, or should we wait for additional feedback?

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:298]: getting confused now, too... if i understand your question correctly: if you want to stop at an exact `n` communities, you can set `cutoff` equal to `best_n` equal to `n`, ie `greedy_modularity_communities(G, cutoff=n, best_n=n)`.

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:296]: This code is just repeated in the while loop.  We *do* need to get communities before the loop, but the rest can all happen in the loop (easier to read).

Also, wherever we check `if len(communities) == 1:` shouldn't that be == cutoff?

Is this loop structure simpler?
```python
communities = next(community_gen)
while len(communities) > cutoff:
    dq = next(community_gen)
    if dq < 0 and len(communities) <= best_n:
        break
    communities = next(community_gen)
```

File: networkx/algorithms/community/modularity_max.py
Comment (dschult) [inline networkx/algorithms/community/modularity_max.py:225]: I guess it is time to shift (or copy) the docstring info that is now in the private generator into this public function.

r"""Find communities in G using greedy modularity maximization.

    This function uses Clauset-Newman-Moore greedy modularity maximization [2]_.

    Greedy modularity maximization begins with each node in its own community
    and joins the pair of communities that most increases modularity until no
    such pair exists. Two keyword arguments adjust the stopping condition.
    `cutoff` is a lower limit on the number of communities so you can stop
    the process before reaching a maximum. `best_n` is an upper limit on the
    number of communities so you can make the process continue until at most
    n communities remain even if the maximum modularity occurs for more.
    To obtain exactly n communities, set both `cutoff` and `best_n` to n.

    This function maximizes the generalized modularity, where `resolution`
    is the resolution parameter, often expressed as $\gamma$.
    See :func:`~networkx.algorithms.community.quality.modularity`.
```

The references and See Also sections should also be copied to the public function.

**Review by dschult — COMMENTED** [review]
more attempts to simplify -- and I think we can turn back to the docstrings again, :}

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:296]: When shifting `dq` into the loop, we never need to check for `len(communities) == 1/cutoff`. that was just a remaining artefact from previous iterations...

**Comment by dschult** [comment]
We have a deprecation question for this PR we wish to replace the keyword `n_communities` with one named `cutoff`.  I believe we need to keep the old keyword, and check its value to see if it is set. If so, raise a deprecation warning, set `cutoff` to the value from `n_communities` and run the function.  If both `cutoff` and `n_communities` are not their default values, then raise an exception. 

Would it be better to not deprecate, and instead change the keyword argument name? Then an unknown keyword argument error is raised upon calling the function with `n_communities` set. The release notes can say which keyword argument to use instead, or more likely the doc_string of the function would make clear that cutoff is the keyword argument to be used.

Which is the better way to proceed?

**Review by dschult — APPROVED** [review]
Feel free to change the doc_strings that I edited to the way you want them. I figured it was easier to put in my edits and have you see them there than to comment on each line separately.  

To load these changes to your local branch, you will need to pull these changes to your local repository from your github repository:   `git pull origin main`  should do the trick. Then edit, commit, push as usual to add new changes.

**Comment by martacki** [comment]
thanks @dschult for support & swift revision!

File: networkx/algorithms/community/modularity_max.py
Comment (rossbar) [inline networkx/algorithms/community/modularity_max.py:261]: The `n_communities` parameter should still be listed here even though it's deprecated. The deprecation can be included in the docstring with the `.. deprecated` directive, e.g. something like:

```rst
n_communities : int

    .. deprecated:: 3.0
       The `n_communities` parameter is deprecated - use `cutoff` and/or
       `best_n` to set bounds on the desired number of communities instead.

    <original description for n_communities here>
```

File: networkx/algorithms/community/tests/test_modularity_max.py
Comment (rossbar) [inline networkx/algorithms/community/tests/test_modularity_max.py:319]:
```suggestion
    # Two additional merging steps:
```

I think? (based on the previous comment)

File: networkx/algorithms/community/modularity_max.py
Comment (rossbar) [inline networkx/algorithms/community/modularity_max.py:56]:
```suggestion
        that leads to the largest modularity.
```
I think, or maybe "will lead" is better - the choice is yours!

File: networkx/algorithms/community/modularity_max.py
Comment (rossbar) [inline networkx/algorithms/community/modularity_max.py:279]: There should be a `Raises` section here to notify users of all the corner cases that cause exceptions, e.g.

```rst
Raises
------
ValueError
    If the `cutoff` or `best_n`  value is not in the range
    ``[1, G.number_of_nodes()]``, or if `best_n` < `cutoff`.
    Also raised if `cutoff` is used with the deprecated `n_communities`
    parameter.

**Review by rossbar — COMMENTED** [review]
The reorganized code + implementation LGTM, though I may have missed some of the design goals/considerations from the lengthy discussion!

I mostly had comments about documentation + grammatical nits, no major blockers. I do think there's a potentially interesting solution to changing up the signature, but that can be pushed to a follow-up issue/PR so it can be discussed separately.

File: networkx/algorithms/community/modularity_max.py
Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:261]: I used the old `n_communities` description as suggested, but isn't it clearer if we copy the `cutoff` docstring? It's supposed to replace `n_communities` and has the same functionality. But I'm fine either way.

Comment (rossbar) [reply networkx/algorithms/community/modularity_max.py:261]: +1 for reusing the `cutoff` description instead if it's more clear

Comment (martacki) [reply networkx/algorithms/community/modularity_max.py:261]: done!

**Review by rossbar — APPROVED** [review]
Excellent, thanks @martacki . A massive thank you for sticking with it to help hammer out the API through so many iterations!

**Pull request merged**
